server/repo-mgr.c

```diff
@@ -4424,3 +4422,90 @@ seaf_repo_manager_convert_repo_path (SeafRepoManager *mgr,
     return ret;
 }
 
+int
+seaf_repo_manager_set_repo_status(SeafRepoManager *mgr,
+                                  const char *repo_id, int status)
+{
+    int ret = 0;
+
+    if (seaf_db_statement_query (mgr->seaf->db,
+                                 "UPDATE RepoInfo SET "
+                                 "status=? WHERE repo_id=?",
+                                 2, "int", status,
+                                 "string", repo_id) < 0)
+        ret = -1;
+
+    return ret;
+}
+
+int
+seaf_repo_manager_get_repo_status(SeafRepoManager *mgr,
+                                  const char *repo_id)
+{
+    char *sql;
+    gboolean is_vir = seaf_repo_manager_is_virtual_repo(mgr, repo_id);
+    if (is_vir)
+        sql = "SELECT status FROM RepoInfo WHERE repo_id="
+              "(SELECT origin_repo FROM VirtualRepo WHERE repo_id=?)";
+    else
+        sql = "SELECT status FROM RepoInfo WHERE repo_id=?";
+
+    return seaf_db_statement_get_int (mgr->seaf->db, sql,
+                                      1, "string", repo_id);
+}
+
+static gboolean
+fill_in_origin_info_cb(SeafDBRow *row, void *data)
+{
+    SeafRepo *repo = NULL;
+    GList **p_list = data;
+
+    const char *name = seaf_db_row_get_column_text (row, 0);
+    int status = seaf_db_row_get_column_int (row, 1);
+
+    if (name != NULL) {
+        repo = g_new0 (SeafRepo, 1);
+        repo->name = g_strdup(name);
+        repo->status = status;
+        *p_list = g_list_prepend (*p_list, repo);
+    }
+
+    return FALSE;
+}
+
+int
+fill_in_origin_info(GList *repos)
+{
```

Comment:
This operation will be too expensive. Adding a lot of database queries for an uncommon case. An easier solution would be to set the repo status for all sub-libraries in set_repo_status(). And also disallow sharing sub-folders in a read-only library. This way no new sub-library will be created after the library is set to read-only status. We also need to disallow creating sub-library too, which will be the case if user sync a sub-folder.